prog/storage/migrate_spdk_vm_to_ubiblk.rb

```diff
@@ -0,0 +1,186 @@
+    end
+    hop_migrate_from_spdk_to_ubiblk
+  end
+
+  label def migrate_from_spdk_to_ubiblk
+    unit_name = "migrate_from_spdk_to_ubiblk_#{vm.inhost_name}"
+    state = vm.vm_host.sshable.d_check(unit_name)
+    case state
+    when "Succeeded"
+      vm.vm_host.sshable.d_clean(unit_name)
+      hop_create_ubiblk_systemd_unit
+    when "NotStarted"
+      vm.vm_host.sshable.d_run(unit_name, "/tmp/migrate", "-base-image=#{base_image_path}", "-overlay-image=#{root_dir_path}disk.raw.bk", "-output-image=#{root_dir_path}disk.raw", "-kek-file=#{kek_file_path}", "-vhost-backend-conf-file=#{vhost_conf_path}")
+      write_kek_pipe
+      nap 5
+    when "InProgress"
+      nap 5
+    else
+      Clog.emit((state == "Failed") ? "could not migrate from spdk to ubiblk" : "got unknown state from daemonizer2 check: #{state}")
+      nap 65536
+    end
+  end
+
+  label def create_ubiblk_systemd_unit
+    vm.vm_host.sshable.cmd("sudo chown #{vm.inhost_name}:#{vm.inhost_name} #{root_dir_path}disk.raw")
+    vm.vm_host.sshable.cmd("sudo host/bin/spdk-migration-helper create-vhost-backend-service-file", stdin: migration_script_params)
+    hop_start_ubiblk_systemd_unit
+  end
+
+  label def start_ubiblk_systemd_unit
+    disk_index = 0
+    unit_name = "#{vm.inhost_name}-#{disk_index}-storage.service"
+    vm.vm_host.sshable.cmd("sudo systemctl enable #{unit_name}")
+    vm.vm_host.sshable.cmd("sudo systemctl start #{unit_name}")
+    write_kek_pipe
+    hop_start_vm
+  end
+
+  label def start_vm
+    vm.vm_host.sshable.cmd("sudo systemctl start #{vm.inhost_name}")
+    hop_update_vm_model
+  end
+
+  label def update_vm_model
+    DB.transaction do
+      vm.strand.update(label: "wait")
+      vm.vm_storage_volumes.first.update(use_bdev_ubi: false, vhost_block_backend_id: vm_host_vhost_block_backend.id, vring_workers: 1, spdk_installation_id: nil)
+    end
+    hop_create_prep_json_file
+  end
+
+  label def create_prep_json_file
+    vm.vm_host.sshable.cmd("sudo tee /vm/#{vm.inhost_name}/prep.json >/dev/null", stdin: vm.params_json)
+    pop "Vm successfully migrated to ubiblk"
+  end
+
+  def migration_script_params
+    {
+      "slice" => vm.vm_host_slice.name,
+      "device" => vm.vm_storage_volumes.first.storage_device.name,
+      "vm_name" => vm.inhost_name,
+      "encrypted" => true,
+      "disk_index" => 0,
+      "vhost_block_backend_version" => Config.vhost_block_backend_version,
+      "max_read_mbytes_per_sec" => vm.vm_storage_volumes.first.max_read_mbytes_per_sec,
+      "max_write_mbytes_per_sec" => vm.vm_storage_volumes.first.max_write_mbytes_per_sec,
+      "spdk_version" => vm.vm_storage_volumes.first.spdk_installation.version
+    }.to_json
+  end
+
+  def root_dir_path
+    "/var/storage/#{vm.inhost_name}/0/"
+  end
+
+  def kek_file_path
+    "#{root_dir_path}kek.pipe"
+  end
+
+  def vhost_conf_path
+    "#{root_dir_path}vhost-backend.conf"
+  end
+
+  def write_kek_pipe
+    vm_storage_volume = vm.vm_storage_volumes.first
```

Comment:
Probably better to do `kek = vm.vm_storage_volumes.first.key_encryption_key_1`, and use that in the rest of the method.